File: src/States.cpp
#include "States.h"

namespace Automata{
    bool States::contains(const State *s) {
        for(auto const& i : _states){
            if(i == s) return true;
            if(*i == *s) return true;
        }
        return false;
    }

    void States::add(State* s) {
        if(!contains(s))
            _states.push_back(s);
    }

    size_t States::size(){
        return _states.size();
    }

    bool States::empty() {
        return size() == 0;
    }

    State* States::operator [](const int& i){
        if(_states.size() >= i || i < 0)
            throw out_of_range(to_string(i));
        return _states.at(i);
    }

    States::States(const int &n) {
        for(int i = 0; i < n; i++){
            State s(n);
            _states.push_back(&s);
        }
    }

    States::States(State* states[], size_t size) {
        for(size_t i = 0; i < size; i++)
            if(!contains(states[i]))
                _states.push_back(states[i]);
    }

    __wrap_iter<vector<State*, std::__1::allocator<State *>>::pointer> States::begin () {
        return _states.begin();
    }

    __wrap_iter<vector<State*, std::__1::allocator<State *>>::pointer> States::end () {
        return _states.end();
    }
    States::operator string () {
        if(_states.empty()) return "{ }";
        string ret = "{ " + _states[0]->getName();

        for(int i = 1; i < _states.size(); i++)
            ret += ", " + _states[i]->getName();

        return ret + " }";
    }
}

File: System.Automata/States.cs
using System.Collections;
using System.Collections.Generic;

namespace System.Automata {
	/// <summary>
	/// The set of all states
	/// </summary>
	public class States : IEnumerable<State> {
		private readonly List<State> _states;

		/// <summary>
		/// Initialize the set with n states name qi for each <![CDATA[i < n]]>.
		/// </summary>
		/// <param name="n">The number of states</param>
		public States(int n) {
			_states = new List<State>(n);
			for(var i = 0; i < n; i++)
				_states.Add(new State(i));
		}
		
		/// <summary>
		/// Initialize the set with the given states.
		/// </summary>
		/// <param name="q">The states to include</param>
		public States(params State[] q) {
			_states = new List<State>(q.Length);
			_states.AddRange(q);
		}
		
		/// <summary>
		/// Add a state to the set.
		/// </summary>
		/// <param name="q">The state to add</param>
		public void Add(State q) {
			if(_states.Exists(x => x.Name == q.Name))
				return;
			_states.Add(q);
		}
		
		/// <summary>
		/// Check if the state exists in the set
		/// </summary>
		/// <param name="q">The state to check</param>
		/// <returns>True if it exists in the set.</returns>
		public bool Contains(State q) {
			return _states.Contains(q);
		}

		public int Count => _states.Count;
		
		public State this[int i] => _states[i];

		public IEnumerator<State> GetEnumerator() {
			return _states.GetEnumerator();
		}

		IEnumerator IEnumerable.GetEnumerator() {
			return GetEnumerator();
		}

		public override string ToString() {
			var ret = "{";
			_states.ForEach(x => ret += $"{x.Name}, ");
			return $"{ret.Substring(0, ret.Length - 2)}}}";
		}
	}
}